### PR TITLE
ref(buffers): Support redis cluster

### PR DESCRIPTION
**Pull request by untitaker**
See https://www.notion.so/sentry/Redis-at-Sentry-Problems-63e322e8b3c74055b927485fc2a13ad8 for some background information on what drives this change.

In principle we want to get rid of redis blaster wherever possible in order to support higher-level abstractions over our use of Redis, but also just to simplify sentry.utils.redis.

Add a way to run buffers on redis cluster.

There is no additional affordance for a migration path, and there is
no guarantee that the data stored in a redis blaster-type cluster can
be read at all in redis-cluster mode. It is expected that buffers is
migrated to a new cluster using ServiceDelegator, i.e. data is
dual-written to both old and new cluster. We have done so before to
migrate between redis hosts.

This was a fairly straightforward migration, there are however two callouts:

* The redis blaster implementation extensively uses `self.cluster.all()` to write the same exact key name to multiple redis shards. This appears to be done just to shard the buffer queue across multiple hosts, but at the same time there is already the concept of "partitions" that does the same exact thing, so the redis-cluster implementation just expects you to use partitions which are then sharded across hosts by redis-cluster-client.

* Since we are doing a storage migration, I am taking this opportunity to get rid of any use of pickle. When connecting via redis blaster, pickle is used, when connecting to a redis cluster, json is used.

*tagging some random people who have dealt with this before AFAIK*

### CODE REVIEW

File: src/sentry/buffer/redis.py [hunk 1]
```diff
@@ -194,10 +200,10 @@ def incr(self, model, columns, filters, extra=None, signal_only=None, return_inc
             # hook here
             # e.g. "update score if last_seen or times_seen is changed"
             for column, value in extra.items():
-                # TODO(dcramer): once this goes live in production, we can kill the pickle path
-                # (this is to ensure a zero downtime deploy where we can transition event processing)
-                pipe.hset(key, "e+" + column, pickle.dumps(value))
```
Comment (wedamija) [inline src/sentry/buffer/redis.py:199]: I have a really vague memory of us trying to kill this pickle before and having it go badly, but I don't remember why.

**Review by wedamija — COMMENTED** [review]
If we dual write won't that end up causing duplicated increments?

I feel like what we might need to do is:
 - Have code reading from both old and new cluster
 - Switch writes to new cluster
 - When we're satisfied buffers on the old cluster are drained, kill old cluster + workers

File: src/sentry/buffer/redis.py (hunk 1 above)
Comment (untitaker) [reply src/sentry/buffer/redis.py:199]: @wedamija going to address both your comments here. IIRC it was @beezz who used servicedelegator to migrate buffers between redis hosts, I didn't think this through but i was going to pick whatever strategy he chose there to do the same kind of migration off of redis blaster towards redis cluster. your plan to dual-*read* and then switch over writes sounds reasonable to me.

the json codepath at least succeeds in tests but i can imagine there is some buffer arg combination out there that does not pass through the json serializer but is pickleable. the only relevant commit i can find is https://github.com/getsentry/sentry/commit/38f65994444b98cb7c6d94029796fa93f6488b86, so perhaps it failed last time just due to the fact that deploys are not atomic and there was no migration to a new cluster, which should not be a problem here.

we'd probably want a percentage-based rollout anyway so we can catch any pickle errors and other problems, eg. regarding load.

File: src/sentry/buffer/redis.py [hunk 2]
```diff
@@ -233,19 +242,32 @@ def process_pending(self, partition=None):
...
-                    if not keys:
-                        continue
-                    keycount += len(keys)
-                    for key in keys:
-                        pending_buffer.append(key.decode("utf-8"))
-                        if pending_buffer.full():
-                            process_incr.apply_async(kwargs={"batch_keys": pending_buffer.flush()})
-                    conn.target([host_id]).zrem(pending_key, *keys)
+            if self.is_redis_cluster:
+                keys = self.cluster.zrange(pending_key, 0, -1)
```
Comment (wedamija) [inline src/sentry/buffer/redis.py:246]: Does this imply that the pending key will now be located on only one node, vs being on multiple nodes in the previous implementation? Are there any concerns around scale there?

**Review by wedamija — COMMENTED** [review]
lgtm in general, one more question

File: src/sentry/buffer/redis.py (hunk 2 above)
Comment (untitaker) [reply src/sentry/buffer/redis.py:246]: yes, i would increase `pending_partitions` to compensate for that

**Comment by beezz** [comment]
> If we dual write won't that end up causing duplicated increments?
> 
> I feel like what we might need to do is:
> 
>     * Have code reading from both old and new cluster
> 
>     * Switch writes to new cluster
> 
>     * When we're satisfied buffers on the old cluster are drained, kill old cluster + workers

For prev migration we had backend that did just the redis commands but not the increaments/db mutations ~ https://github.com/getsentry/getsentry/pull/8421

**Comment by untitaker** [comment]
I'd ask yall to take another look. While the migration strategy isn't figured out yet, I am in contact with @beezz about it. Here are some notes from our last meeting:

1. michal believes we should refactor buffers so multiple instances can be run in the same process. this would allow us to migrate individual usecases to a new cluster vs all usecases at once
2. `partitions` as replacement for having the same key on all nodes is something we just have to try out. there's the potential that the additional tasks that are spawned as part of increased `partitions` will cause trouble, but that's something to just test

File: src/sentry/buffer/redis.py [hunk 3]
```diff
@@ -172,20 +175,23 @@ def incr(self, model, columns, filters, extra=None, signal_only=None, return_inc
...
 
         pipe = conn.pipeline()
         pipe.hsetnx(key, "m", f"{model.__module__}.{model.__name__}")
-        # TODO(dcramer): once this goes live in production, we can kill the pickle path
-        # (this is to ensure a zero downtime deploy where we can transition event processing)
-        pipe.hsetnx(key, "f", pickle.dumps(filters))
-        # pipe.hsetnx(key, 'f', json.dumps(self._dump_values(filters)))
+
+        if self.is_redis_cluster:
+            pipe.hsetnx(key, "f", json.dumps(self._dump_values(filters)))
```
Comment (mitsuhiko) [inline src/sentry/buffer/redis.py:191]: Let's figure out a path here to validate the shape of the JSON data even for the non JSON path so we know that we have a path forward.

**Review by mitsuhiko — CHANGES_REQUESTED** [review]

**Comment by untitaker** [comment]
Reopening this PR, but before we do that I'm going to do some validation that JSON is viable to switch to

prs related to that:

* https://github.com/getsentry/sentry/pull/51137
* https://github.com/getsentry/sentry/pull/51142
* https://github.com/getsentry/sentry/pull/51146
* https://github.com/getsentry/getsentry/pull/10879

**Comment by untitaker** [comment]
@mitsuhiko i believe this PR is ready to merge. I am still validating that all of our arguments are pickleable before we cut over, but this shouldn't prevent this implementation from landing.

**Review by mitsuhiko — APPROVED** [review]
Let's try to ship it

**Pull request merged**